List and describe post-MVP features in more detail (#114)
WebAssembly/gc: proposals/gc/Post-MVP.md

# GC Post-v1 Extensions

> This document discusses various extensions which seem desirable for comprehensive support of GC types, but are intentionally left out of the [MVP](MVP.md), in order to keep its scope manageable.

Reviewer: Perhaps add a sentence to the effect of: "Over the course of implementing and validating the MVP, it is possible that features in this list may be promoted to the MVP if experience shows that MVP performance would not otherwise be viable." ?

Author: Done.

To that end, _bulk copying_ instructions could be added, similar to the [bulk instructions for tables and memories](https://github.com/WebAssembly/bulk-memory-operations).

**Why Post-MVP:** These operators do not provide any additional expressiveness, so are not essential for the MVP.

### Sketch

* An instruction for bulk copying a struct:
  - `struct.copy $d : [(ref $d) (ref $s)] -> []` where both `$d` and `$s` are struct types, `$d` has only mutable fields, and `$s <: $d` modulo mutability

* An instruction for bulk copying an array range:
  - `array.copy $d : [(ref $d) i32 (ref $s) i32 i32] -> []`
    - iff both `$d` and `$s` are array types
    - and `$d` has mutable element type
    - and `$s <: $d` modulo mutability
  - the remaining operands are destination and source offset and length of the range

* An instruction for bulk setting an array range:
  - `array.fill $d : [(ref $d) i32 t i32] -> []`
    - iff `$d = (array (mut st))`
    - and `t = unpacked(st)`
  - the remaining operands are destination offset, initialisation value, and length of the range

## "Butterfly Objects"

> One common suggestion is to merge structs and arrays into a single construct with both struct-like fields and a array-like elements.

Author: Good point, renamed to "arrays with fields".

### Sketch

* Add a new form of statically-sized array type:
  - `arraytype ::= ... | array N <fieldtype>`

* Such an array type is a subtype of a smaller statically-sized array:
  - `array N1 ft1 <: array N2 ft2`
    - iff `N1 < N2`
    - and `ft1 <: ft2`

* Such an array type also is a subtype of a dynamically-sized array:
  - `array N ft1 <: array ft2`
    - iff `ft1 <: ft2`

* An instruction for allocating a statically-sized array:
  - `array.new_static $t : [t^N] -> [(ref $t)]`
    - iff `$t = (array N ft)`
    - and `ft = (mut? st)`
    - and `t = unpacked(st)`

## Nested Data Structures

The MVP only supports "flat" data structures, i.e., structs or arrays whose field types are simple values.
Ultimately, Wasm should support more of the C data model, where structs and arrays can be nested in an unboxed fashion, i.e., flattened into a single heap object.

With the extension sketched below, it is possible, for example, to represent the equivalent of ["butterfly objects"](#butterfly-objects), by nesting a dynamically-sized array at the end of the struct.
For example:
```
(type $Array (array i32))
(type $Butterfly (struct f32 i64 (type $Array))

> ```

Author: Renamed.

* A data type is _fixed_ if it is a struct or array that is not flexible.

* With nesting and flexible aggregates, the type grammar generalizes as follows:
  ```
  datatype        ::=  <fix_datatype> | <flex_datatype>
  fix_datatype    ::=  (struct <fix_fieldtype>*) | (array N <fix_fieldtype>)
  flex_datatype   ::=  (struct <fix_fieldtype>* <flex_fieldtype>) | (array <fix_fieldtype>)

  fix_fieldtype   ::=  (<mut> <storagetype>) | <fix_datatype>
  flex_fieldtype  ::=  <flex_datatype>
  ```
  However, additional constraints apply to (mutually) recursive type definitions in order to ensure well-foundedness of the recursion (a data type cannot contain itself in flat form).
  For example,
  ```
  (type $t (struct (type $t)))
  ```
  is not valid.
  For example, well-foundedness can be ensured by requiring that the *nesting depth* of any `datatype`, derivable by the following inductive definition, is finite:
  ```
  |(<mut> <storagetype>)|  = 0
  |(struct <fieldtype>*)|  = 1 + max{|<fieldtype>|*}
  |(array N? <fieldtype>)| = 1 + |<fieldtype>|
  ```

Allocation:

> * Allocation instructions like `struct.new` expect initialiser operands for nested aggregates: each individual field for a nested struct, and one initialiser for a nested array (which may itself be a list of initialisers, if the array's element type is a again a struct). Details TBD.

Reviewer: ```suggestion
* Allocation instructions like `struct.new` expect initialiser operands for nested aggregates: each individual field for a nested struct, and one initialiser for a nested array (which may itself be a list of initialisers, if the array's element type is again a struct). Details TBD.
```

Author: Done.

  - `struct.inner $t i : [([in]ref null? $t)] -> [(inref ft)]`
    - iff `$t = (struct ft1^i ft ft2*)`
  - `array.inner $t : [([in]ref null? $t) i32] -> [(inref ft)]`
    - iff `$t = (array N? ft)`
  - the source reference may either be a regular or itself interior
  - traps if the source reference is null

  For example:
  ```
  (struct.get $point $y (struct.inner $colored-point $p (<some colored point>)))
  ```

* Instructions to obtain an inner reference from a [field reference](#field-references):
  - `struct.inner_ref : [([in]ref null? $t) (fieldref null? $t ft)] -> [(inref ft)]`
    - iff `$t = (struct ft1^* ft ft2*)`
    - and `ft = (type $t')`
  - traps if either of the references is null

* It is not valid to get or set a field or element that has aggregate type.
  Writing to a nested structure or array requires combined uses of `struct.inner`/`array.inner` to acquire the interior reference and `struct.set`/`array.set` to its contents:
  ```
  (struct.set $color-point $x
    (struct.inner $color-point $p (...some $color-point...))
    (f64.const 1.2)
  )
  ```

  An engine should be able to optimise away intermediate interior pointers very easily.

> * TBD: As sketched here, interior references can only point ot nested aggregates. Should there also be interior references to plain fields?

Reviewer: ```suggestion
* TBD: As sketched here, interior references can only point to nested aggregates. Should there also be interior references to plain fields?
```

Author: Done.

      (local.get $x) (local.gt $x)
    )
  )
  ```

## Variants

The MVP supports the most basic form of _pointer tagging_ via the `i31ref` type.
That type allows injecting and distinguishing unboxed integers and pointers in the same type space, such that unboxing can be guaranteed on all platforms.

However, many language implementations use more elaborate tagging schemes in one form or the other. For example, they want to efficiently distinguish different classes of values without dereferencing, or to store additional information in pointer values without requiring extra space (e.g., Lisp or Prolog often introduce a special tag bit for cons cells).
Other languages provide user-defined tags as an explicit language feature (e.g., _variants_ or _algebraic data types_).

Unfortunately, hardware differs widely in how many tagging bits a pointer can conveniently support, and different VMs might have additional constraints on how many of these bits they can make available to user code.
In order to provide tagging in a way that is portable but maximally efficient on any given hardware and engine, a somewhat higher level of abstraction is useful.

> Such a more high-level solution would be to support a form of _variant types_ (a.k.a. disjoint unions or sum types) in the type system.

Reviewer: ```suggestion
Such a higher-level solution would be to support a form of _variant types_ (a.k.a. disjoint unions or sum types) in the type system.
```

Author: Done.

In addition to structs and arrays, a module could define a variant type that is a closed union of multiple different cases.
Dedicated instructions allow allocating and inspecting references of variant type.

It is left to the engine to pick an efficient representation for the required tags, and depending on the hardware's word size, the number of tags in a defined type, the presence of parameters to a given tag, and other design decisions in the engine, these tags could either be stored as bits in the pointer, in a shared per-type data structure (a.k. hidden class or shape), or in an explicit per-value slot within the heap object.
These decisions can be made by the engine on a per-type basis; validation ensures that all uses are coherent.

> **Why Post-MVP:** Variants allow for more compact representations and potentially more precise types, thereby possibly saving a cetain number of runtime checks and downcasts over the use of a common supertype, as necessary in the MVP. However, they are not strictly necessary in the presence of the latter. Nor can they replace it, since they necessarily define a _closed_ set of values, whereas the ability to import an arbitrary number of abstract types requires a way to include an _open_ set of types. To handle that, yet another form of _extensible union types_ (with generative tags) would be required in addition.

Reviewer 2: ```suggestion
**Why Post-MVP:** Variants allow for more compact representations and potentially more precise types, thereby possibly saving a certain number of runtime checks and downcasts over the use of a common supertype, as necessary in the MVP. However, they are not strictly necessary in the presence of the latter. Nor can they replace it, since they necessarily define a _closed_ set of values, whereas the ability to import an arbitrary number of abstract types requires a way to include an _open_ set of types. To handle that, yet another form of _extensible union types_ (with generative tags) would be required in addition.
```

Reviewer 2: The above change just changes "cetain" to "certain."

  - i.e., if the field type is a scalar, pass its value to the label, otherwise an interior reference
  - TBD: the typing rule could synthesise a more precise result type without the tested case; alternatively, this could be replaced with a multi-branch instruction, but that may be cumbersome in many cases

* TBD: how this integrates with RTTs

## Static Fields

In various object models and value representations, heap values share certain meta information -- for example, the method table in an object, the tag in a union type, or other "static" meta information about a value.

Since a Wasm engine already has to store its own meta information in heap values, such as GC type descriptor or RTTs, that may double the space usage for meta data in every heap object (e.g., two memory words instead of just one). It would hence be desirable if GC type definitions could piggy-back on the meta object that the engine already has to implement.

The basic idea would be introducing a notion of _static fields_ in a form of immutable meta object that is shared between multiple instances of the same type.
There are various ways in which this could be modelled, details are TBD.

> **Why Post-MVP:** Such a feature only saves space, so isn't critical for the MVP. Furthermore, there isn't much precendence for exposing such a mechanism too user code in low-level form, so no obvious design philosophy to follow.

Reviewer 2: ```suggestion
**Why Post-MVP:** Such a feature only saves space, so isn't critical for the MVP. Furthermore, there isn't much precedent for exposing such a mechanism to user code in low-level form, so no obvious design philosophy to follow.
```
precedence => precedent, too => to
